File: app/src/main/kotlin/com/absinthe/libchecker/utils/DownloadUtils.java
package com.absinthe.libchecker.utils;

import android.content.Context;

import androidx.annotation.NonNull;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.ResponseBody;

public class DownloadUtils {

    private volatile static DownloadUtils sDownloadUtil;
    private final OkHttpClient mOkHttpClient;

    private DownloadUtils() {
        mOkHttpClient = new OkHttpClient();
    }

    public static DownloadUtils get() {
        if (sDownloadUtil == null) {
            synchronized (DownloadUtils.class) {
                if (sDownloadUtil == null) {
                    sDownloadUtil = new DownloadUtils();
                }
            }
        }
        return sDownloadUtil;
    }

    /**
     * @param url      Download URL
     * @param file     File
     * @param listener Download callback
     */
    public void download(final Context context, final String url, final File file, final OnDownloadListener listener) {
        Request request = new Request.Builder().url(url).build();
        mOkHttpClient.newCall(request).enqueue(new Callback() {
            @Override
            public void onFailure(@NonNull Call call, @NonNull IOException e) {
                listener.onDownloadFailed();
            }

            @Override
            public void onResponse(@NonNull Call call, @NonNull Response response) throws IOException {
                InputStream is = null;
                byte[] buf = new byte[2048];
                int len;
                FileOutputStream fos = null;
                if (file.exists()) {
                    file.delete();
                }
                file.createNewFile();

                try {
                    ResponseBody body = response.body();
                    if (body != null) {
                        is = body.byteStream();
                        long total = body.contentLength();

                        fos = new FileOutputStream(file);
                        long sum = 0;

                        while ((len = is.read(buf)) != -1) {
                            fos.write(buf, 0, len);
                            sum += len;
                            int progress = (int) (sum * 1.0f / total * 100);
                            listener.onDownloading(progress);
                        }
                        fos.flush();
                        listener.onDownloadSuccess();
                    } else {
                        listener.onDownloadFailed();
                    }
                } catch (Exception e) {
                    listener.onDownloadFailed();
                } finally {
                    try {
                        if (is != null) {
                            is.close();
                        }
                        if (fos != null) {
                            fos.close();
                        }
                    } catch (IOException ignore) {
                    }
                }
            }
        });
    }

    public interface OnDownloadListener {
        void onDownloadSuccess();

        void onDownloading(int progress);

        void onDownloadFailed();
    }
}

File: app/src/main/kotlin/com/absinthe/libchecker/utils/DownloadUtils.kt
package com.absinthe.libchecker.utils

import com.absinthe.libchecker.compat.DnsCompat
import java.io.File
import java.io.IOException
import okhttp3.Call
import okhttp3.Callback
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import org.apache.commons.io.IOUtils

object DownloadUtils {
  private val client by lazy {
    OkHttpClient.Builder()
      .apply {
        // Enables Encrypted Client Hello where the platform and network security config allow it.
        DnsCompat.echCapableDns?.let(::dns)
      }
      .build()
  }

  /**
   * @param url      Download URL
   * @param file     File
   * @param listener Download callback
   */
  fun download(url: String, file: File, listener: OnDownloadListener, maximumBytes: Long = 32L * 1024 * 1024) {
    val request: Request = Request.Builder()
      .url(url)
      .build()
    client.newCall(request).enqueue(object : Callback {
      override fun onFailure(call: Call, e: IOException) {
        listener.onDownloadFailed()
      }

      @Throws(IOException::class)
      override fun onResponse(call: Call, response: Response) {
        val success = runCatching {
          response.use {
            check(it.isSuccessful) { "Download failed: HTTP ${it.code}" }
            check(it.body.contentLength() <= maximumBytes) { "Download is too large" }
            file.parentFile?.mkdirs()
            it.body.byteStream().use { input ->
              file.outputStream().use { output ->
                val limit = maximumBytes.coerceAtLeast(0)
                val count = IOUtils.copyLarge(input, output, 0, limit)
                check(count < limit || input.read() == -1) { "Download is too large" }
                output.fd.sync()
              }
            }
          }
        }.isSuccess
        if (success) {
          listener.onDownloadSuccess()
        } else {
          file.delete()
          listener.onDownloadFailed()
        }
      }
    })
  }

  interface OnDownloadListener {
    fun onDownloadSuccess()
    fun onDownloadFailed()
  }
}
